src/mbgl/gl/attribute.cpp

```diff
@@ -7,41 +7,14 @@ namespace gl {
 
 using namespace platform;
 
-void bindAttributeLocation(Context& context, ProgramID id, AttributeLocation location, const char* name) {
-    // We're using sequentially numberered attribute locations starting with 0. Therefore we can use
-    // the location as a proxy for the number of attributes.
-    if (location >= context.maximumVertexBindingCount) {
```

Comment:
Do we still have an equivalent check in place?

Reply:
The check is done [before render](https://github.com/mapbox/mapbox-gl-native/blob/858097102d9f586e17b118532213e38262e801db/src/mbgl/renderer/render_layer.cpp#L76).

Reply:
@astojilj Correct me if I'm wrong, but the [check](https://github.com/mapbox/mapbox-gl-native/blob/858097102d9f586e17b118532213e38262e801db/src/mbgl/renderer/render_layer.cpp#L76) you mention only takes care of logging, it doesn't prevent rendering. So the queryLocations is still reached. What happens when an attribute can't be found or we cross the `MAX_VERTEX_ATTRIBUTES` limit? If `glGetAttribLocation` returns -1 in that situation, things will go wrong when using the returned location as an index [here](https://github.com/mapbox/mapbox-gl-native/blob/858097102d9f586e17b118532213e38262e801db/src/mbgl/gl/attribute.hpp#L88) no?

Reply:
No, if glGetAttribLocation returns -1, then `if (location) {` [here](https://github.com/mapbox/mapbox-gl-native/blob/858097102d9f586e17b118532213e38262e801db/src/mbgl/gl/attribute.hpp#L88), as false, guards the `result.at(*location) = binding;`.

Binding, or not binding, vertex attributes before linking, based on GL_MAX_VERTEX_ATTRIBS count, doesn't change the behavior of following glLinkProgram and [error](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glLinkProgram.xhtml).
As I saw, this was resolved by [packing attributes PR](https://github.com/mapbox/mapbox-gl-native/issues/8183) and now we use the log to inform about error.